http-server-netty/src/main/java/io/micronaut/http/server/netty/FormRouteCompleter.java

```diff
@@ -0,0 +1,285 @@
+    private static final Logger LOG = LoggerFactory.getLogger(FormRouteCompleter.class);
+
+    private final NettyStreamingFileUpload.Factory fileUploadFactory;
+    private final ConversionService conversionService;
+    private final boolean alwaysAddContent = request.isFormData();
+    private final AtomicLong pressureRequested = new AtomicLong();
+    private final Map<String, Sinks.Many<Object>> subjectsByDataName = new HashMap<>();
+    private final Collection<Sinks.Many<?>> downstreamSubscribers = new ArrayList<>();
+
+    FormRouteCompleter(NettyStreamingFileUpload.Factory fileUploadFactory, ConversionService conversionService, NettyHttpRequest<?> request, RouteMatch<?> routeMatch) {
+        super(request, routeMatch);
+        this.fileUploadFactory = fileUploadFactory;
+        this.conversionService = conversionService;
+    }
+
+    private void request(long n) {
+        pressureRequested.getAndUpdate(old -> {
+            if ((old + n) < old) {
+                return Long.MAX_VALUE;
+            } else {
+                return old + n;
+            }
+        });
+        needsInput = true;
+        Runnable checkDemand = this.checkDemand;
+        if (checkDemand != null) {
+            checkDemand.run();
+        }
+    }
+
+    private <T> Flux<T> withFlowControl(Flux<T> flux, MicronautHttpData<?> data) {
+        return flux
+            .doOnComplete(data::release)
+            .doOnRequest(this::request);
+    }
+
+    @Override
+    protected void addHolder(ByteBufHolder holder) {
+        if (holder instanceof HttpData data) {
+            needsInput = pressureRequested.decrementAndGet() > 0;
+            addData((MicronautHttpData<?>) data);
+        } else {
+            super.addHolder(holder);
+        }
+    }
+
+    @Override
+    void completeSuccess() {
+        for (Sinks.Many<?> subject : downstreamSubscribers) {
+            // subjects will ignore the onComplete if they're already done
+            subject.tryEmitComplete();
+        }
+        super.completeSuccess();
+    }
+
+    @Override
+    void completeFailure(Throwable failure) {
+        super.completeFailure(failure);
+        for (Sinks.Many<?> subject : downstreamSubscribers) {
+            subject.tryEmitError(failure);
+        }
+    }
+
+    private void addData(MicronautHttpData<?> data) {
+        if (LOG.isTraceEnabled()) {
+            LOG.trace("Received HTTP Data for request [{}]: {}", request, data);
+        }
+
+        String name = data.getName();
+        Optional<Argument<?>> requiredInput = routeMatch.getRequiredInput(name);
+
+        if (requiredInput.isEmpty()) {
+            request.addContent(data);
+            request(1);
+            return;
+        }
+
+        Argument<?> argument = requiredInput.get();
+        Supplier<Object> value;
+        boolean isPublisher = Publishers.isConvertibleToPublisher(argument.getType());
+        boolean chunkedProcessing = false;
+
+        if (isPublisher) {
```

Comment:
I think we can because we can detect it fits somewhere. IMHO the correct behavior would be to publish the error in the publisher if the part is not detected or something, assuming it will be used to produce a response.

Reply:
it would be inconsistent with other parameter fulfillment, but maybe thats fine.

i don't want to touch it in this pr though.